R/aes.R

```diff
@@ -131,40 +131,40 @@ new_aesthetic <- function(x, env = globalenv()) {
 }
 
 #' @export
-# TODO: should convert to proper S7 method once bug in S7 is resolved
-`print.ggplot2::mapping` <- function(x, ...) {
-  cat("Aesthetic mapping: \n")
+local({
+  S7::method(print, class_mapping) <- function(x, ...) {
+    cat("Aesthetic mapping: \n")
 
-  if (length(x) == 0) {
-    cat("<empty>\n")
-  } else {
-    values <- vapply(x, quo_label, character(1))
-    bullets <- paste0("* ", format(paste0("`", names(x), "`")), " -> ", values, "\n")
+    if (length(x) == 0) {
+      cat("<empty>\n")
+    } else {
+      values <- vapply(x, quo_label, character(1))
+      bullets <- paste0("* ", format(paste0("`", names(x), "`")), " -> ", values, "\n")
 
-    cat(bullets, sep = "")
+      cat(bullets, sep = "")
+    }
+
+    invisible(x)
   }
+})
 
-  invisible(x)
-}
+local({
+  S7::method(`[`, class_mapping) <- function(x, i, ...) {
+    class_mapping(`[`(S7::S7_data(x), i, ...))
+  }
+})
```

Comment:
Do we expect this to be necessary in perpetuity? If not, maybe add a note about fixing it once the bug in S7 has been fixed